tft.py

```diff
@@ -0,0 +1,213 @@
+from cdragontoolbox.binfile import BinFile, BinHashBase, BinHashValue
+from json import JSONEncoder
+import json
+import mmap
+import copy
+import re
+import os
+
+
+class NaiveJsonEncoder(JSONEncoder):
+    def default(self, other):
+        if isinstance(other, BinHashBase):
+            if other.s is None:
+                return other.hex
+            return other.s
+        return other.__dict__
+
+
+class Transformer:
+
+    output_prefix = r"D:\pbe\tft_temp"
+    prefix = r"D:\pbe"
+    map22_file = os.path.join(prefix, r"data\maps\shipping\map22\map22.bin")
+    font_folder = os.path.join(prefix, r"data\menu")
+    character_folder = os.path.join(prefix, r"data\characters")
+    exported_file = os.path.join(prefix, r"files.exported.txt")
+
+    def __init__(self):
+        with open(self.exported_file, "rb", 0) as f:
+            with mmap.mmap(f.fileno(), 0, access=mmap.ACCESS_READ) as s:
+                self.description_files = re.findall(br"game/data/menu/(fontconfig_.._..\.txt)", s)
+
+        map22 = BinFile(self.map22_file)
+
+        self.traits = self.parse_traits(map22)
+        self.champs = self.parse_champs(map22, self.traits)
+        self.items = self.parse_items(map22)
+        lang_files = self.description_files
+
+        # clean up data for JSON export
+        [x.pop("internal") for x in self.traits]
+        [self.items[x].pop("internal") for x in self.items]
+        [self.items[x].pop("newf") for x in self.items]
+        del self.description_files
+
+        os.mkdir(self.output_prefix)
+        template = json.dumps(self, cls=NaiveJsonEncoder, indent=4, sort_keys=True)
+        with open(os.path.join(self.output_prefix, "template.json"), "w+", encoding="utf-8") as f:
+            f.write(template)
+
+        for language in lang_files:
+            use = copy.deepcopy(template)
+            code = str(language, "utf-8")[11:-4]
+
+            font_file = os.path.join(self.font_folder, str(language, "utf-8"))
+            with open(font_file, "rb", 0) as f:
+                with mmap.mmap(f.fileno(), 0, access=mmap.ACCESS_READ) as s:
+                    for line in iter(s.readline, b""):
+                        line = str(line, "utf-8")
+                        if line.startswith("tr"):
+                            split = line.split("=")
+                            key = split[0][4:-2]
+                            val = split[1][2:-3]
+                            use = use.replace(key, val)
```

Comment:
If I'm not mistaking, you "blindly" replace strings into JSON encoded data, not checking anything (not even word boundaries). This is definitely not clean.

Reply:
the keys in this case are unique enough that it works, but if you have a better idea, im all for it